plugin/pkg/global/resourcereferencemanager/admission.go

```diff
@@ -319,6 +322,68 @@ func (r *ReferenceManager) Admit(ctx context.Context, a admission.Attributes, o
 				})
 			}
 		}
+	case core.Kind("CloudProfile"):
+		cloudProfile, ok := a.GetObject().(*core.CloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into CloudProfile object")
+		}
+		if utils.SkipVerification(operation, cloudProfile.ObjectMeta) {
+			return nil
+		}
+		if a.GetOperation() == admission.Update {
+			oldCloudProfile, ok := a.GetOldObject().(*core.CloudProfile)
+			if !ok {
+				return apierrors.NewBadRequest("could not convert old resource into CloudProfile object")
+			}
+
+			// getting Kubernetes versions that have been removed from the CloudProfile
+			removedKubernetesVersions := getRemovedVersions(oldCloudProfile.Spec.Kubernetes.Versions, cloudProfile.Spec.Kubernetes.Versions)
+
+			// getting Machine image versions that have been removed from the CloudProfile
+			removedMachineImageVersions := map[string]sets.String{}
+			for _, oldImage := range oldCloudProfile.Spec.MachineImages {
+				imageFound := false
+				for _, newImage := range cloudProfile.Spec.MachineImages {
+					if oldImage.Name == newImage.Name {
+						imageFound = true
+						removedMachineImageVersions[oldImage.Name] = getRemovedVersions(oldImage.Versions, newImage.Versions)
+					}
+				}
+				if !imageFound {
+					for _, version := range oldImage.Versions {
+						if removedMachineImageVersions[oldImage.Name] == nil {
+							removedMachineImageVersions[oldImage.Name] = sets.NewString()
+						}
+						removedMachineImageVersions[oldImage.Name] = removedMachineImageVersions[oldImage.Name].Insert(version.Version)
+					}
+				}
+			}
+
+			if len(removedKubernetesVersions) > 0 || len(removedMachineImageVersions) > 0 {
+				shootList, e := r.shootLister.List(labels.Everything())
+				if e != nil {
+					return err
+				}
+				for _, shoot := range shootList {
+					if shoot.Spec.CloudProfileName != cloudProfile.Name {
+						continue
+					}
+
+					if removedKubernetesVersions.Has(shoot.Spec.Kubernetes.Version) {
+						err = multierror.Append(err, fmt.Errorf("unable to delete Kubernetes version '%s' from CloudProfile. Version is still in use by a Shoot (Name: '%s', Namespace: '%s', CloudProfile: '%s')", shoot.Spec.Kubernetes.Version, shoot.Name, shoot.Namespace, shoot.Spec.CloudProfileName))
+					}
+					for _, worker := range shoot.Spec.Provider.Workers {
+						// happens if Shoot already runs a version that does not exist in the CloudProfile - in this case: ignore
+						if _, ok := removedMachineImageVersions[worker.Machine.Image.Name]; !ok {
```

Comment:
`worker.Machine.Image` is a pointer - `nil` check is missing

Reply:
right, the defaulting of the ShootMachineImage  in the shootAdmissionPlugin happens after - thanks!